core/src/Instance/B2CAuthority.cs

```diff
@@ -37,37 +38,85 @@ internal class B2CAuthority : AadAuthority
+        public const string MicrosoftOnline = "https://login.microsoftonline.com";
+        public const string OpenIdConfigurationEndpoint = "v2.0/.well-known/openid-configuration";
+        internal const string B2CTrustedHost = "b2clogin.com";
 
-        internal B2CAuthority(string authority, bool validateAuthority) 
+        internal B2CAuthority(string authority, bool validateAuthority)
             : base(authority, validateAuthority)
         {
+            AuthorityType = AuthorityType.B2C;
+            ValidateAuthority = validateAuthority;
+
             Uri authorityUri = new Uri(authority);
-            string[] pathSegments = authorityUri.AbsolutePath.Substring(1).Split(new [] { '/'}, StringSplitOptions.RemoveEmptyEntries);
+            string[] pathSegments = authorityUri.AbsolutePath.Substring(1).Split(new[] { '/' }, StringSplitOptions.RemoveEmptyEntries);
             if (pathSegments.Length < 3)
             {
                 throw new ArgumentException(CoreErrorMessages.B2cAuthorityUriInvalidPath);
             }
 
             CanonicalAuthority = string.Format(CultureInfo.InvariantCulture, B2CCanonicalAuthorityTemplate, authorityUri.Authority,
                 pathSegments[0], pathSegments[1], pathSegments[2]);
-
-            AuthorityType = AuthorityType.B2C;
         }
 
-        protected override async Task<string> GetOpenIdConfigurationEndpointAsync(
-            IHttpManager httpManager, 
-            ITelemetryManager telemetryManager, 
-            string userPrincipalName, 
+        internal override async Task UpdateCanonicalAuthorityAsync(
+            IHttpManager httpManager,
+            ITelemetryManager telemetryManager,
             RequestContext requestContext)
         {
             if (ValidateAuthority && !IsInTrustedHostList(new Uri(CanonicalAuthority).Host))
             {
                 throw new ArgumentException(CoreErrorMessages.UnsupportedAuthorityValidation);
             }
+            
+            var metadata = await AadInstanceDiscovery
+                             .Instance.GetMetadataEntryAsync(
+                                 httpManager,
+                                 telemetryManager,
+                                 new Uri(UpdateCanonicalAuthorityB2C()),
+                                 ValidateAuthority,
+                                 requestContext)
+                             .ConfigureAwait(false);
+
+            CanonicalAuthority = UpdateHost(CanonicalAuthority, metadata.PreferredNetwork);
+        }
+
+        private string UpdateCanonicalAuthorityB2C()
+        {
+            Uri b2cAuthority = new Uri(CanonicalAuthority);
+
+            string canonicalAuthorityUri = string.Format(CultureInfo.InvariantCulture, MicrosoftOnline + b2cAuthority.AbsolutePath);
+            return canonicalAuthorityUri;
+        }
 
+        internal new static bool IsInTrustedHostList(string host)
```

Comment:
>internal new static [](start = 8, length = 19)

not a fan of this construct. I suggest adding an overload or enabling passing in an additional delegate to the IsInTrustedHostlist, alternatively given the code is called from here it's possible to simply create a new method.

Reply:
made a new method instead.

---
In reply to: [230835339](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1383#discussion_r230835339) [](ancestors = 230835339)